sdk/messaging/azeventhubs/checkpoints/blob_store.go

```diff
@@ -94,48 +109,80 @@ func (b *BlobStore) ListCheckpoints(ctx context.Context, fullyQualifiedNamespace
 		return nil, err
 	}
 
-	pager := b.cc.NewListBlobsFlatPager(&container.ListBlobsFlatOptions{
-		Prefix: &prefix,
-		Include: container.ListBlobsInclude{
-			Metadata: true,
-		},
-	})
+	correctPrefix := strings.ToLower(legacyPrefix)
 
-	var checkpoints []azeventhubs.Checkpoint
-
-	for pager.More() {
-		resp, err := pager.NextPage(ctx)
+	checkpointsMap := map[string]azeventhubs.Checkpoint{}
 
-		if err != nil {
-			return nil, err
-		}
+	addCheckpointsToMap := func(prefix string, isLegacy bool) error {
+		pager := b.cc.NewListBlobsFlatPager(&container.ListBlobsFlatOptions{
+			Prefix: &prefix,
+			Include: container.ListBlobsInclude{
+				Metadata: true,
+			},
+		})
 
-		for _, blob := range resp.Segment.BlobItems {
-			partitionID := partitionIDRegexp.FindString(*blob.Name)
+		for pager.More() {
+			resp, err := pager.NextPage(ctx)
 
-			cp := azeventhubs.Checkpoint{
-				FullyQualifiedNamespace: fullyQualifiedNamespace,
-				EventHubName:            eventHubName,
-				ConsumerGroup:           consumerGroup,
-				PartitionID:             partitionID,
+			if err != nil {
+				return err
 			}
 
-			if err := updateCheckpoint(blob.Metadata, &cp); err != nil {
-				return nil, err
+			for _, blob := range resp.Segment.BlobItems {
+				partitionID := partitionIDRegexp.FindString(*blob.Name)
+
+				cp := azeventhubs.Checkpoint{
+					FullyQualifiedNamespace: fullyQualifiedNamespace,
+					EventHubName:            eventHubName,
+					ConsumerGroup:           consumerGroup,
+					PartitionID:             partitionID,
+				}
+
+				if err := copyCheckpointPropsFromMetadata(blob.Metadata, &cp); err != nil {
+					return err
+				}
+
+				if !isLegacy {
+					cp.ConsumerGroup = strings.ToLower(cp.ConsumerGroup)
+					cp.EventHubName = strings.ToLower(cp.EventHubName)
+					cp.FullyQualifiedNamespace = strings.ToLower(cp.FullyQualifiedNamespace)
+					cp.PartitionID = strings.ToLower(cp.PartitionID)
+				}
+
+				checkpointsMap[strings.ToLower(*blob.Name)] = cp
 			}
+		}
+
+		return nil
+	}
 
-			checkpoints = append(checkpoints, cp)
+	if correctPrefix != legacyPrefix {
+		// we've got a casing difference - we need to make sure we look at both possible
+		// casings because of our previous bug where we didn't ToLower() the blob name by
+		// default.
+		if err := addCheckpointsToMap(legacyPrefix, true); err != nil {
+			return nil, err
 		}
 	}
 
+	if err := addCheckpointsToMap(correctPrefix, false); err != nil {
```

Comment:
You could condense this into `addCheckpointsToMap(correctPrefix, correctPrefix != legacyPrefix)` and remove the previous block.

Reply:
Ugh maybe not.